### ISSUE TITLE
Scheduled jobs are only executed when server restarts

### ISSUE THREAD

**Issue by amangeot** [agenda/agenda#848]
Hello,
First of all, thank you for this project, it's awesome. I have an issue with jobs being fired only after server restarts. Do you see any mistake in the way I use the api?

Here is how they are scheduled:
```
  agenda.schedule(when, name, data)
```

And its job definition:
```
agenda.define(name, async function (job, done) {
  // Some async tasks
  if (done) done()
})
```

Here is how the instance is initiated:
```
const agenda = new Agenda({
  db: {
    address: config.agenda.mongoDbAddress,
    collection: config.agenda.collection,
  },
})

agenda.on('ready', function () {
  agenda.start()
})
```

Any help would be greatly appreciated,

**Comment by amangeot** [comment]
It seems to run just fine in development environment.

Agenda: v1.0.3
Node: v6.11.2
Npm: v5.2.0
On Windows 10

**Comment by dmbarreiro** [comment]
Hello @amangeot ,
the only thing I see for now is that you have a typo on this line `await someAsyncTasks(jobs.attrs.data)`, it should be `job.attrs.data` instead of `jobs.attrs.data`. That's probably not the issue but better to be safe than sorry.

So if I understood correctly you suffer the problem only in production but not in development. Could you check if Node time in sync with what you expect in the problematic environment?

When you say jobs are fired only after server restarts you mean the old jobs that were not triggered before or you mean new jobs introduces after the restart or both?

This toy example I've built based on your code works on my system (macOS, node v12.4.0, npm 6.11.3):

```
const express = require("express");
const Agenda = require("agenda");

const toPromise = function(param) {
  return new Promise(function(resolve, rejects) {
    console.log(param);
    resolve(true);
  });
};

const agenda = new Agenda({
  db: {
    address: 'mongodb://127.0.0.1/agenda',
    collection: "testAgenda"
  }
});

agenda.define("test", async function(job, done) {
  const result = await toPromise(job.attrs.data);
  console.log(result);
  if (done) done();
});

const app = express();
const server = app.listen();

agenda.on("ready", async function() {
  await agenda.start();
  await agenda.schedule("in 10 seconds", "test", { to: "admin@example.com" });
});
```

**Comment by amangeot** [comment]
Hello @dmbarreiro , thank you for taking the time to look at it.

> When you say jobs are fired only after server restarts you mean the old jobs that were not triggered before or you mean new jobs introduces after the restart or both?
Once server is started, jobs are properly scheduled and saved into the database, though they wont execute at the time they should. When I restart the server, all jobs that should have ran earlier (but did not) are executed at the same time (or about).

I was wondering if it could be some kind of circular dependencies but I'm not sure why they would happen in production only.

Here is my project structure:
```
~/agenda/index.js
  => defines and exports an instance of agenda = new Agenda(..), sets agenda.on('ready', ...)
~/routers/bookings/agenda/constants.js
 => exports constant strings for jobs' name
~/routers/bookings/agenda/jobs.js
  => imports agenda instance from '~/agenda/index.js, job names from './constants.js' and calls agenda.define(name, func)
~/routers/bookings/routers/index.js
  => imports agenda instance from '~/agenda/index.js', job names, and calls some agenda.schedule(..) at appropriate times
~/routers/bookings/index.js
  => export sub-router and require('./agenda/jobs) where jobs are defined
```
And similar structures with subrouters at `/routers/partners/..`

I saw you had a project structure where you import functions calling agenda.define(..) passing agenda instance as a param. Could this be the issue?

**Comment by dmbarreiro** [comment]
Yeah, it is possible. What you see of having all the jobs scheduled executed at once is what agenda does when an scheduled task is overdue, it runs it immediately. So, when that happens (agenda runs all your ovedue tasks after restart), if you schedule new taks, are they run at the expected time or also never executed?

It could be a wrong or missing import somewhere.
Make sure your defines exist when you schedule the task. You can try as a test to export agenda on `~/routers/bookings/agenda/jobs.js` and import that agenda on `~/routers/bookings/routers/index.js` to make sure the defines are there when you shedule the task. Could be that you are scheduling tasks that are not defined and when you execute again the tasks are defined and overdue so they are al executed immediately but as you said I don't know why would that happen only in production.

About what you mentioned of calling `agenda.define(...)` passing agenda instance as a param I've never done it like that, the agenda instance is the one you are calling the method on and checking the `define` method definition I don't see an option for that, I haven't been using agenda for such a long time though so maybe it is possible but anyway it should work the way you use it now.

**Comment by amangeot** [comment]
Hello thanks for the feedback, 
I believe it's fixed, complete misusage on my end, sorry.

I meant this project structure where you pass `agenda` to a function that will call `agenda.define(..)`:
https://github.com/agenda/agenda#example-project-structure I'll dig in

Looking at MongoDB's collections, I figured out some jobs were never "done", they probably lock the scheduler for other jobs:

I'm not saving job IDs / cancelling them when necessary, I'm using stop conditions to prevent actions within job definitions but I was not calling the "done" param in one of them. This probably jammed the job queue.

This should fix the issue:
```
  if (/* some conditions */) {
    if (done) done() // WNG: Do not omit this in agenda.define(..)
    return
  }
```

Thank you very much for your time, sorry about my mistake, awesome library!
